packages/flutter/lib/src/widgets/async.dart

```diff
@@ -542,30 +657,68 @@ class FutureBuilder<T> extends StatefulWidget {
+  ///    future has previously completed, in which case the previous result
+  ///    persists.
+  ///
+  ///    If this widget was created without initial data via
+  ///    [new FutureBuilder.withoutInitialData], then the [AsyncSnapshot.data]
+  ///    will be unset, and attempts to access the data will result in an
+  ///    exception.
+  ///
+  ///  * [ConnectionState.waiting]: [future] is not null but has not yet
+  ///    completed.
   ///
-  ///  * [ConnectionState.waiting]: [future] is not null, but has not yet
-  ///    completed. The [AsyncSnapshot.data] will be set to [initialData],
-  ///    unless a future has previously completed, in which case the previous
-  ///    result persists.
+  ///    If this widget was created with initial data via [new FutureBuilder],
+  ///    then the [AsyncSnapshot.data] will be set to [initialData], unless a
+  ///    future has previously completed, in which case the previous result
+  ///    persists.
+  ///
+  ///    If this widget was created without initial data via
+  ///    [new FutureBuilder.withoutInitialData], then the [AsyncSnapshot.data]
+  ///    will be unset, and attempts to access the data will result in an
+  ///    exception.
   ///
   ///  * [ConnectionState.done]: [future] is not null, and has completed. If the
   ///    future completed successfully, the [AsyncSnapshot.data] will be set to
   ///    the value to which the future completed. If it completed with an error,
   ///    [AsyncSnapshot.hasError] will be true and [AsyncSnapshot.error] will be
   ///    set to the error object.
+  ///
+  ///    In the case of [future] being a [Future<void>] or a [Future<Null>],
+  ///    the snapshot will not contain data even if the future completed
+  ///    successfully.
   final AsyncWidgetBuilder<T> builder;
 
+  /// Whether this builder's initial snapshot contains data.
+  ///
+  /// If this is false, then attempting to access [initialData] will throw an
+  /// error.
+  ///
+  /// See also:
+  ///
+  ///  * [AsyncSnapshot.hasData], the corresponding property that will be set
+  ///    in the initial snapshot.
+  final bool hasInitialData;
+
   /// The data that will be used to create the snapshots provided until a
   /// non-null [future] has completed.
   ///
-  /// If the future completes with an error, the data in the [AsyncSnapshot]
-  /// provided to the [builder] will become null, regardless of [initialData].
-  /// (The error itself will be available in [AsyncSnapshot.error], and
+  /// If the future completes with an error, the [AsyncSnapshot] provided to
+  /// the [builder] will contain no data, regardless of [initialData]. (The
+  /// error itself will be available in [AsyncSnapshot.error], and
   /// [AsyncSnapshot.hasError] will be true.)
-  final T initialData;
+  ///
+  /// If [hasInitialData] is false, this will throw a [StateError].
+  T get initialData {
+    if (!hasInitialData)
+      throw StateError('No initial data');
```

Comment:
as above, assuming we keep this feature at all. it's weird to me to have "initial data" for a future.